crates/engine/tree/src/tree/invalid_block_hook.rs

```diff
@@ -49,3 +49,31 @@ impl InvalidBlockHook for NoopInvalidBlockHook {
     ) {
     }
 }
+
+/// A chain of [`InvalidBlockHook`]s that are executed in order.
+pub struct InvalidBlockHookChain(pub Vec<Box<dyn InvalidBlockHook>>);
+
+impl std::fmt::Debug for InvalidBlockHookChain {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("InvalidBlockHookChain").field("len", &self.0.len()).finish()
+    }
+}
+
+impl InvalidBlockHook for InvalidBlockHookChain {
+    fn on_invalid_block(
+        &self,
+        block: SealedBlockWithSenders,
+        header: SealedHeader,
+        output: BlockExecutionOutput<Receipt>,
```

Comment:
should we make all of these & and expect the hooks to clone on demand?

Reply:
agreed